traits/trait_set_object.py

```diff
@@ -8,251 +8,448 @@
+            notifier(removed, added)
+
+    # ------------------------------------------------------------------------
+    # set interface
+    # ------------------------------------------------------------------------
 
     def __deepcopy__(self, memo):
+        """ Perform a deepcopy operation.
+
+        Notifiers are transient and should not be copied.
+        """
         id_self = id(self)
         if id_self in memo:
             return memo[id_self]
 
-        memo[id_self] = result = TraitSetObject(
-            self.trait,
-            lambda: None,
-            self.name,
+        # notifiers are transient and should not be copied
+        memo[id_self] = result = TraitSet(
             [copy.deepcopy(x, memo) for x in self],
+            validator=copy.deepcopy(self.validator, memo),
+            notifiers=[],
         )
 
         return result
 
+    def __getstate__(self):
+        """ Get the state of the object for serialization.
+
+        Notifiers are transient and should not be serialized.
+        """
+        result = self.__dict__.copy()
+        # notifiers are transient and should not be serialized
+        result.pop("notifiers", None)
+        return result
+
+    def __setstate__(self, state):
+        """ Restore the state of the object after serialization.
+
+        Notifiers are transient and are restored to the empty list.
+        """
+        state['notifiers'] = []
+        self.__dict__.update(state)
+
+    def add(self, value):
+        """ Add an element to a set.
+
+        This has no effect if the element is already present.
+
+        Parameters
+        ----------
+        value : any
+            The value to add to the set.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                An empty set.
+            added : set
+                A set containing the added value.
+
+        """
+        validated_values = self.validate(value)
+        if len(validated_values) > 1:
+            raise ValueError("Validator returned {} values "
+                             "where 1 value is "
+                             "expected".format(len(validated_values)))
+
+        value = next(iter(validated_values))
+        if value not in self:
+            super().add(value)
+            self.notify(set(), validated_values)
+
+    def remove(self, value):
+        """ Remove an element from a set; it must be a member.
+
+        If the element is not a member, raise a KeyError.
+
+        Parameters
+        ----------
+        value : any
+            An element in the set
+
+        Raises
+        ------
+        KeyError
+            If the value is not found in the set.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                A set containing the removed value.
+            added : set
+                An empty set.
+
+        """
+        validated_values = self.validate(value)
```

Comment:
The validator should only be given added values.
See #978

Reply:
Merged #978